service/matching/db_task_ownership.go

```diff
@@ -0,0 +1,334 @@
+// The MIT License
+//
+// Copyright (c) 2020 Temporal Technologies Inc.  All rights reserved.
+//
+// Copyright (c) 2020 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package matching
+
+import (
+	"fmt"
+	"sync"
+	"time"
+
+	enumspb "go.temporal.io/api/enums/v1"
+	"go.temporal.io/api/serviceerror"
+
+	persistencespb "go.temporal.io/server/api/persistence/v1"
+	"go.temporal.io/server/common/clock"
+	"go.temporal.io/server/common/log"
+	"go.temporal.io/server/common/persistence"
+	"go.temporal.io/server/common/primitives/timestamp"
+)
+
+//go:generate mockgen -copyright_file ../../LICENSE -package $GOPACKAGE -source $GOFILE -destination db_task_ownership_mock.go
+
+const (
+	dbTaskOwnershipStatusUninitialized dbTaskOwnershipStatus = 0
+	dbTaskOwnershipStatusOwning        dbTaskOwnershipStatus = 1
+	dbTaskOwnershipStatusLost          dbTaskOwnershipStatus = 2
+)
+
+type (
+	dbTaskOwnershipStatus int
+
+	dbTaskOwnership interface {
+		shutdownChan() <-chan struct{}
+		getAckedTaskID() int64
+		updateAckedTaskID(taskID int64)
+		getLastAllocatedTaskID() int64
+		persistTaskQueue() error
+		flushTasks(taskInfos ...*persistencespb.TaskInfo) error
+	}
+
+	dbTaskOwnershipState struct {
+		rangeID             int64
+		ackedTaskID         int64
+		lastAllocatedTaskID int64
+		minTaskID           int64 // exclusive
```

Comment:
maybe make the inclusivity part of the name so user doesn't have to come here to look it up (i.e. minTaskIdExclusive)?